src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/RemoteTagHelperResolver.cs

```diff
@@ -2,89 +2,70 @@
-    private readonly CompilationTagHelperResolver _compilationTagHelperResolver;
-
-    public RemoteTagHelperResolver(ITelemetryReporter telemetryReporter)
-    {
-        _fallbackFactory = new FallbackProjectEngineFactory();
-        _compilationTagHelperResolver = new CompilationTagHelperResolver(telemetryReporter);
-    }
-
-    public ValueTask<TagHelperResolutionResult> GetTagHelpersAsync(
-        Project workspaceProject,
-        IProjectSnapshot projectSnapshot,
-        CancellationToken cancellationToken)
-        => throw new NotImplementedException();
+    private readonly IFallbackProjectEngineFactory _fallbackFactory = new FallbackProjectEngineFactory();
+    private readonly Dictionary<string, IProjectEngineFactory> _typeNameToFactoryMap = new(StringComparer.Ordinal);
+    private readonly CompilationTagHelperResolver _compilationTagHelperResolver = new(telemetryReporter);
 
     public ValueTask<TagHelperResolutionResult> GetTagHelpersAsync(
         Project workspaceProject,
         RazorConfiguration? configuration,
         string factoryTypeName,
         CancellationToken cancellationToken)
     {
-        if (workspaceProject is null)
-        {
-            throw new ArgumentNullException(nameof(workspaceProject));
-        }
-
         if (configuration is null)
         {
             return new(TagHelperResolutionResult.Empty);
         }
 
-        var projectEngine = CreateProjectEngine(configuration, factoryTypeName);
-
-        return _compilationTagHelperResolver.GetTagHelpersAsync(workspaceProject, projectEngine, cancellationToken);
+        return _compilationTagHelperResolver.GetTagHelpersAsync(
+            workspaceProject,
+            CreateProjectEngine(configuration, factoryTypeName),
+            cancellationToken);
     }
 
-    private RazorProjectEngine CreateProjectEngine(RazorConfiguration? configuration, string factoryTypeName)
+    private RazorProjectEngine CreateProjectEngine(RazorConfiguration configuration, string factoryTypeName)
     {
-        // This section is really similar to the code DefaultProjectEngineFactoryService
-        // but with a few differences that are significant in the remote scenario
-        //
-        // Most notably, we are going to find the Tag Helpers using a compilation, and we have
-        // no editor settings.
-        //
-        // The default configuration currently matches MVC-2.0. Beyond MVC-2.0 we added SDK support for
-        // properly detecting project versions, so that's a good version to assume when we can't find a
-        // configuration.
-        configuration ??= s_defaultConfiguration;
-
         // If there's no factory to handle the configuration then fall back to a very basic configuration.
         //
         // This will stop a crash from happening in this case (misconfigured project), but will still make
         // it obvious to the user that something is wrong.
```

Comment:
Unrelated, but I love the optimism of this comment, that users will report "I say good fellow, my Razor editor appears to be using a fallback configuration" rather than just "intellisense is broken".

Reply:
I agree! I'm not in love with the whole fallback project engine thing, but I was amused that the deleted code above never would have been hit because the only thing calling this method checked `configuration` for null.